### PR TITLE
Fix dependence of Uniform logp on bound method

### LINKED ISSUE #4499: ValueError when using check_bounds=False in a model with DiscreteUniform
I was poking around to see if we should change the docstrings of check_bounds in pm.Model to mention this might not be safe in the presence of discrete variables and I found this strange error:

```python
with pm.Model(check_bounds=False) as m:
    y = pm.DiscreteUniform('y', 0, 5)
    trace = pm.sample()
```
```
Traceback (most recent call last):
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/IPython/core/interactiveshell.py", line 3418, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-60-ca8c3478b1dc>", line 3, in <module>
    trace = pm.sample()
  File "/home/ricardo/Documents/Projects/pymc3/pymc3/sampling.py", line 515, in sample
    step = assign_step_methods(model, step, step_kwargs=kwargs)
  File "/home/ricardo/Documents/Projects/pymc3/pymc3/sampling.py", line 217, in assign_step_methods
    return instantiate_steppers(model, steps, selected_steps, step_kwargs)
  File "/home/ricardo/Documents/Projects/pymc3/pymc3/sampling.py", line 138, in instantiate_steppers
    step = step_class(vars=vars, **args)
  File "/home/ricardo/Documents/Projects/pymc3/pymc3/step_methods/metropolis.py", line 187, in __init__
    self.delta_logp = delta_logp(model.logpt, vars, shared)
  File "/home/ricardo/Documents/Projects/pymc3/pymc3/step_methods/metropolis.py", line 888, in delta_logp
    logp1 = pm.CallableTensor(logp0)(inarray1)
  File "/home/ricardo/Documents/Projects/pymc3/pymc3/aesaraf.py", line 308, in __call__
    (oldinput,) = inputvars(self.tensor)
ValueError: not enough values to unpack (expected 1, got 0)
```

I don't see why changing the check_bounds flag would lead to any issues in sampling. I can call the logp method for any value and I get a finite (and wrong) evaluation as would be expected from removing the bounds:

```python
np.array([m.logp({'y': y}) for y in np.arange(-10, 10)])
array([-1.79175947, -1.79175947, -1.79175947, -1.79175947, -1.79175947,
       -1.79175947, -1.79175947, -1.79175947, -1.79175947, -1.79175947,
       -1.79175947, -1.79175947, -1.79175947, -1.79175947, -1.79175947,
       -1.79175947, -1.79175947, -1.79175947, -1.79175947, -1.79175947])
```

I tried with a couple of other Discrete Variables and I had no issues either. Obviously it works as expected with the flag set to True

```python
with pm.Model(check_bounds=True) as m:
    y = pm.DiscreteUniform('y', 0, 5)
    trace = pm.sample()
```

### PR DESCRIPTION
**Pull request by ricardoV94**
This PR fixes #4499  

The Uniform and Discrete Uniform logp methods depended on the dist_math::bound method for correct evaluation of multiple values. This is problematic because the bound can be disabled by setting `check_bounds = False` in `pm.Model()`.

Other changes:
1. I tried to improve the documentation of the bound method to emphasize this. I also tried to make it a bit more clear what the "broadcast_conditions" secret keyword argument means and when it should be used. I would suggest this to be made an optional keyword argument in the function call of `bound`, and perhaps even rename it to `multivariate=False` to make it more obvious when should it be used. Let me know if you agree and I can add it to this PR as well.

2. I tweaked the docstring of `check_bounds` to emphasize this should not be used for models containing discrete variables as some of these depend on bound checks when sampled (unless they have specific samplers that cannot propose invalid values or the logp happens to behave nicely when given out-of-bound points)

3. Removed rendudant `tt.all()` calls in the bound conditions of several multivariate distributions.

*** 

Do we need a release note?

Depending on what your PR does, here are a few things you might want to address in the description:
+ [x] ~~what are the (breaking) changes that this PR makes?~~ None
+ [x] ~~important background, or details about the implementation~~ None
+ [x] are the changes—especially new features—covered by tests and docstrings? Tests cover known bugs. Docstrings improved
+ [x] [linting/style checks have been run](https://github.com/pymc-devs/pymc3/wiki/PyMC3-Python-Code-Style)
+ [x] ~~[consider adding/updating relevant example notebooks](https://github.com/pymc-devs/pymc-examples)~~
+ [x] right before it's ready to merge, mention the PR in the RELEASE-NOTES.md

### CODE REVIEW

**Review by twiecki — APPROVED** [review]

**Comment by twiecki** [comment]
LGTM - thanks! Still needs a line in the release notes.

File: RELEASE-NOTES.md [hunk 1]
```diff
@@ -13,6 +13,8 @@
 - The `pymc3.memoize` module was removed and replaced with `cachetools`.  The `hashable` function and `WithMemoization` class were moved to `pymc3.util` (see [#4509](https://github.com/pymc-devs/pymc3/pull/4509)).
 - Remove float128 dtype support (see [#4514](https://github.com/pymc-devs/pymc3/pull/4514)).
 - `pm.make_shared_replacements` now retains broadcasting information which fixes issues with Metropolis samplers (see [#4492](https://github.com/pymc-devs/pymc3/pull/4492)).
+- Logp method of `Uniform` and `DiscreteUniform` no longer depends on `pymc3.distributions.dist_math.bound` for proper evaluation (see [#4541](https://github.com/pymc-devs/pymc3/pull/4541)).
```
Comment (twiecki) [inline RELEASE-NOTES.md:13]: Something is wrong here as we already release `3.11.2` this should go into `vNext`.

Comment (ricardoV94) [reply RELEASE-NOTES.md:13]: Right, it seems the release notes were only changed in V3. Should I rebase the release notes from here? https://github.com/pymc-devs/pymc3/blob/v3/RELEASE-NOTES.md

Comment (michaelosthege) [reply RELEASE-NOTES.md:13]: Good question. The Release notes are the only document that spans branches and will continue to even after `v4` was released.
I don't think we'll be able to keep up with rebasing it going into the future.

Can you bundle these items in a `vNext (4.0.0)` section, but still point out those that were already released in `3.11.2`? You can link to the Release itself, where the release notes are duplicated.

Comment (ricardoV94) [reply RELEASE-NOTES.md:13]: Sure, let me give it a try

File: RELEASE-NOTES.md [hunk 2]
```diff
@@ -1,19 +1,37 @@
...
+- The `CAR` distribution has been added to allow for use of conditional autoregressions which often are used in spatial and network models.
+- ...
 
 ### Maintenance
-- The `pymc3.memoize` module was removed and replaced with `cachetools`.  The `hashable` function and `WithMemoization` class were moved to `pymc3.util` (see [#4509](https://github.com/pymc-devs/pymc3/pull/4509)).
 - Remove float128 dtype support (see [#4514](https://github.com/pymc-devs/pymc3/pull/4514)).
+- Logp method of `Uniform` and `DiscreteUniform` no longer depends on `pymc3.distributions.dist_math.bound` for proper evaluation (see [#4541](https://github.com/pymc-devs/pymc3/pull/4541)).
+- ...
+
+## PyMC3 3.11.2 (14 March 2021)
```
Comment (michaelosthege) [inline RELEASE-NOTES.md:16]: I think we shouldn't include the entire section: Some changes in `v3` (the stats and plotting functions) are not included on the master branch. See #4528.

Instead, we could just mark the "also-in-v3-changes" with "(released in `3.11.2`)"?

Comment (michaelosthege) [reply RELEASE-NOTES.md:16]: @MarcoGorelli do you know of best practices when it comes to changelog files in a world with multiple branches and backporting?

Comment (ricardoV94) [reply RELEASE-NOTES.md:16]: I don't know. What about just repeating again this line from V3.11.0?

> - ⚠ Many plotting and diagnostic functions that were just aliasing ArviZ functions were removed (see [4397](https://github.com/pymc-devs/pymc3/pull/4397/files#)). This includes `pm.summary`, `pm.traceplot`, `pm.ess` and many more!

Comment (ricardoV94) [reply RELEASE-NOTES.md:16]: That's the only bump between V3.11.1 -> V3.11.2 -> V4.0.0 no?

Comment (michaelosthege) [reply RELEASE-NOTES.md:16]: For now we could do that. I'll make sure to do the PR such that `pm.plots` and `pm.stats` are re-introduced on `master` too.
Then we can edit the release notes again..

**Pull request merged**